File: way-to-fit-backend/src/main/kotlin/com/waytofit/competition/adapter/in/web/dto/LeaderboardDto.kt
package com.waytofit.competition.adapter.`in`.web.dto

import com.waytofit.competition.application.port.`in`.EventLeaderboardResult
import com.waytofit.competition.application.port.`in`.LeaderboardEntry
import com.waytofit.competition.application.port.`in`.OverallLeaderboardResult
import com.waytofit.competition.application.port.`in`.OverallLeaderboardEntry
import com.waytofit.competition.domain.enums.ResultStatus
import java.math.BigDecimal
import java.util.UUID

data class EventLeaderboardResponse(
    val eventId: UUID,
    val entries: List<LeaderboardEntryResponse>,
) {
    companion object {
        fun fromResult(result: EventLeaderboardResult) = EventLeaderboardResponse(
            eventId = result.eventId,
            entries = result.entries.map { LeaderboardEntryResponse.fromEntry(it) }
        )
    }
}

data class OverallLeaderboardResponse(
    val stageId: UUID,
    val entries: List<OverallLeaderboardEntryResponse>,
) {
    companion object {
        fun fromResult(result: OverallLeaderboardResult) = OverallLeaderboardResponse(
            stageId = result.stageId,
            entries = result.entries.map { OverallLeaderboardEntryResponse.fromEntry(it) }
        )
    }
}

data class LeaderboardEntryResponse(
    val rank: Int,
    val registrationId: UUID,
    val participantName: String,
    val scaleCategory: String,
    val resultTimeSeconds: Int?,
    val resultRounds: Int?,
    val resultReps: Int?,
    val resultWeight: BigDecimal?,
    val resultCustom: String?,
    val resultStatus: ResultStatus?,
    val videoUrl: String,
    val memberIds: List<UUID>,
) {
    companion object {
        fun fromEntry(entry: LeaderboardEntry) = LeaderboardEntryResponse(
            rank = entry.rank,
            registrationId = entry.registrationId,
            participantName = entry.participantName,
            scaleCategory = entry.scaleCategory,
            resultTimeSeconds = entry.resultTimeSeconds,
            resultRounds = entry.resultRounds,
            resultReps = entry.resultReps,
            resultWeight = entry.resultWeight,
            resultCustom = entry.resultCustom,
            resultStatus = entry.resultStatus,
            videoUrl = entry.videoUrl,
            memberIds = entry.memberIds
        )
    }
}

data class OverallLeaderboardEntryResponse(
    val rank: Int,
    val registrationId: UUID,
    val participantName: String,
    val scaleCategory: String,
    val totalPoints: Int,
    val eventRanks: Map<UUID, Int>,
    val manualRank: Int?,
    val memberIds: List<UUID>,
) {
    companion object {
        fun fromEntry(entry: OverallLeaderboardEntry) = OverallLeaderboardEntryResponse(
            rank = entry.rank,
            registrationId = entry.registrationId,
            participantName = entry.participantName,
            scaleCategory = entry.scaleCategory,
            totalPoints = entry.totalPoints,
            eventRanks = entry.eventRanks,
            manualRank = entry.manualRank,
            memberIds = entry.memberIds
        )
    }
}

data class OverrideRankRequest(
    val manualRank: Int?,
) {
    fun toCommand(registrationId: UUID) = com.waytofit.competition.application.port.`in`.OverrideRankCommand(
        registrationId = registrationId,
        manualRank = manualRank
    )
}

File: way-to-fit-app/lib/features/competition/data/dto/leaderboard_dto.dart
import 'package:json_annotation/json_annotation.dart';

import '../../domain/models.dart';

part 'leaderboard_dto.g.dart';

@JsonSerializable()
class EventLeaderboardResponseDto {
  const EventLeaderboardResponseDto({
    required this.eventId,
    required this.entries,
  });

  final String eventId;
  final List<LeaderboardEntryResponseDto> entries;

  factory EventLeaderboardResponseDto.fromJson(Map<String, dynamic> json) =>
      _$EventLeaderboardResponseDtoFromJson(json);

  Map<String, dynamic> toJson() => _$EventLeaderboardResponseDtoToJson(this);
}

@JsonSerializable()
class OverallLeaderboardResponseDto {
  const OverallLeaderboardResponseDto({
    required this.stageId,
    required this.entries,
  });

  final String stageId;
  final List<OverallLeaderboardEntryResponseDto> entries;

  factory OverallLeaderboardResponseDto.fromJson(Map<String, dynamic> json) =>
      _$OverallLeaderboardResponseDtoFromJson(json);

  Map<String, dynamic> toJson() => _$OverallLeaderboardResponseDtoToJson(this);
}

@JsonSerializable()
class LeaderboardEntryResponseDto {
  const LeaderboardEntryResponseDto({
    required this.rank,
    required this.registrationId,
    required this.participantName,
    required this.scaleCategory,
    required this.resultTimeSeconds,
    required this.resultRounds,
    required this.resultReps,
    required this.resultWeight,
    required this.resultCustom,
    required this.resultStatus,
    required this.videoUrl,
    required this.memberIds,
  });

  final int rank;
  final String registrationId;
  final String participantName;
  final String scaleCategory;
  final int? resultTimeSeconds;
  final int? resultRounds;
  final int? resultReps;
  final num? resultWeight;
  final String? resultCustom;
  final String? resultStatus;
  final String videoUrl;
  final List<String> memberIds;

  factory LeaderboardEntryResponseDto.fromJson(Map<String, dynamic> json) =>
      _$LeaderboardEntryResponseDtoFromJson(json);

  Map<String, dynamic> toJson() => _$LeaderboardEntryResponseDtoToJson(this);

  LeaderboardEntry toDomain() => LeaderboardEntry(
    rank: rank,
    registrationId: registrationId,
    participantName: participantName,
    scaleCategory: scaleCategory,
    resultTimeSeconds: resultTimeSeconds,
    resultRounds: resultRounds,
    resultReps: resultReps,
    resultWeight: resultWeight,
    resultCustom: resultCustom,
    resultStatus: resultStatus,
    videoUrl: videoUrl,
    memberIds: memberIds,
  );
}

@JsonSerializable()
class OverallLeaderboardEntryResponseDto {
  const OverallLeaderboardEntryResponseDto({
    required this.rank,
    required this.registrationId,
    required this.participantName,
    required this.scaleCategory,
    required this.totalPoints,
    required this.eventRanks,
    required this.manualRank,
    required this.memberIds,
  });

  final int rank;
  final String registrationId;
  final String participantName;
  final String scaleCategory;
  final int totalPoints;
  final Map<String, int> eventRanks;
  final int? manualRank;
  final List<String> memberIds;

  factory OverallLeaderboardEntryResponseDto.fromJson(
    Map<String, dynamic> json,
  ) => _$OverallLeaderboardEntryResponseDtoFromJson(json);

  Map<String, dynamic> toJson() =>
      _$OverallLeaderboardEntryResponseDtoToJson(this);

  OverallLeaderboardEntry toDomain() => OverallLeaderboardEntry(
    rank: rank,
    registrationId: registrationId,
    participantName: participantName,
    scaleCategory: scaleCategory,
    totalPoints: totalPoints,
    eventRanks: eventRanks,
    manualRank: manualRank,
    memberIds: memberIds,
  );
}
